File: app/src/main/java/com/adriaan/claudeusage/data/api/ClaudeApiService.kt
package com.adriaan.claudeusage.data.api

import com.adriaan.claudeusage.data.model.AccountInfo
import com.adriaan.claudeusage.data.model.Organization
import com.adriaan.claudeusage.data.model.UsageData
import com.adriaan.claudeusage.data.model.UsageLimit
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.logging.HttpLoggingInterceptor
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

class ClaudeApiService(private val cookieString: String) {

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .addInterceptor { chain ->
            val request = chain.request().newBuilder()
                .addHeader("Cookie", cookieString)
                .addHeader("User-Agent", "Mozilla/5.0 (Linux; Android 14) AppleWebKit/537.36")
                .addHeader("Accept", "application/json")
                .addHeader("Referer", "https://claude.ai/")
                .addHeader("Origin", "https://claude.ai")
                .build()
            chain.proceed(request)
        }
        .addInterceptor(HttpLoggingInterceptor().apply {
            level = HttpLoggingInterceptor.Level.BODY
        })
        .build()

    /** HTTP status code + body (body null on transport failure). status -1 = exception. */
    private data class HttpResult(val status: Int, val body: String?)

    private fun get(path: String): HttpResult {
        return try {
            val request = Request.Builder()
                .url("https://claude.ai$path")
                .get()
                .build()
            val response = client.newCall(request).execute()
            val body = response.body?.string()
            android.util.Log.d("ClaudeUsageNet", "GET $path -> ${response.code} (${body?.length ?: 0} bytes)")
            HttpResult(response.code, if (response.isSuccessful) body else null)
        } catch (e: Exception) {
            android.util.Log.w("ClaudeUsageNet", "GET $path failed: ${e.message}")
            HttpResult(-1, null)
        }
    }

    fun getOrganizations(): List<Organization> {
        val json = get("/api/organizations").body ?: return emptyList()
        return try {
            val arr = JSONArray(json)
            (0 until arr.length()).map { i ->
                val obj = arr.getJSONObject(i)
                Organization(
                    uuid = obj.optString("uuid"),
                    id = obj.optString("id"),
                    name = obj.optString("name"),
                    planTier = obj.optString("plan_tier").ifEmpty { null },
                    billingType = obj.optString("billing_type").ifEmpty { null }
                )
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    /**
     * Result of a usage fetch. On success [limits] is non-empty and [raw] holds the JSON that
     * produced it. On failure [limits] is empty and [raw] holds a diagnostic of every endpoint
     * tried with its HTTP status — surfaced in the UI so a moved/auth-failed endpoint is visible
     * instead of silently showing zeros.
     */
    data class UsageResult(val limits: List<UsageLimit>, val raw: String?)

    fun getUsage(orgId: String): UsageResult {
        // claude.ai serves Pro usage from /usage as percent-per-window limit objects.
        // Keep fallbacks in case the path moves.
        val endpoints = listOf(
            "/api/organizations/$orgId/usage",
            "/api/organizations/$orgId/limits",
            "/api/account/usage",
            "/api/usage"
        )

        val diagnostics = StringBuilder("No usage endpoint returned usable data.\n\nTried:\n")
        for (endpoint in endpoints) {
            val result = get(endpoint)
            val statusLabel = if (result.status == -1) "network error" else "HTTP ${result.status}"
            diagnostics.append("• $endpoint → $statusLabel\n")

            val raw = result.body ?: continue
            val limits = UsageData.parseLimits(raw)
            if (limits.isNotEmpty()) return UsageResult(limits, raw)
        }
        return UsageResult(emptyList(), diagnostics.toString())
    }

    fun getAccountInfo(): AccountInfo? {
        val endpoints = listOf("/api/auth/session", "/api/account", "/api/me")
        for (endpoint in endpoints) {
            val raw = get(endpoint).body ?: continue
            try {
                val obj = JSONObject(raw)
                val user = obj.optJSONObject("user") ?: obj
                return AccountInfo(
                    id = user.optString("id"),
                    uuid = user.optString("uuid"),
                    email = user.optString("email"),
                    name = user.optString("name").ifEmpty { null },
                    displayName = user.optString("display_name").ifEmpty { null }
                )
            } catch (e: Exception) {
                continue
            }
        }
        return null
    }
}

File: ios/Shared/ClaudeAPIService.swift
import Foundation

/// Talks to claude.ai's internal, undocumented API using the session cookie captured at login.
///
/// This mirrors the Android `ClaudeApiService`: it adds the `Cookie` header to every request,
/// lists organizations, then tries a handful of usage endpoints and parses the first that returns
/// usable limit windows.
struct ClaudeAPIService {

    let cookieString: String

    private static let base = "https://claude.ai"

    private var session: URLSession {
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 15
        config.httpCookieStorage = nil          // we set the Cookie header ourselves, like Android
        config.httpShouldSetCookies = false
        return URLSession(configuration: config)
    }

    /// HTTP status + body. `status == -1` means a transport/exception failure. `body` is nil
    /// unless the response was 2xx (so callers never parse an error page).
    private struct HTTPResult { let status: Int; let body: String? }

    private func get(_ path: String) async -> HTTPResult {
        guard let url = URL(string: Self.base + path) else { return HTTPResult(status: -1, body: nil) }
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue(cookieString, forHTTPHeaderField: "Cookie")
        request.setValue("Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15",
                         forHTTPHeaderField: "User-Agent")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue(Self.base + "/", forHTTPHeaderField: "Referer")
        request.setValue(Self.base, forHTTPHeaderField: "Origin")

        do {
            let (data, response) = try await session.data(for: request)
            let http = response as? HTTPURLResponse
            let code = http?.statusCode ?? -1
            let body = String(data: data, encoding: .utf8)
            let ok = (200...299).contains(code)
            return HTTPResult(status: code, body: ok ? body : nil)
        } catch {
            return HTTPResult(status: -1, body: nil)
        }
    }

    // MARK: Organizations

    func getOrganizations() async -> [Organization] {
        guard let body = await get("/api/organizations").body,
              let data = body.data(using: .utf8),
              let arr = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return []
        }
        return arr.map { o in
            Organization(
                uuid: o["uuid"] as? String ?? "",
                id: o["id"] as? String ?? "",
                name: o["name"] as? String ?? "",
                planTier: (o["plan_tier"] as? String).flatMap { $0.isEmpty ? nil : $0 },
                billingType: (o["billing_type"] as? String).flatMap { $0.isEmpty ? nil : $0 }
            )
        }
    }

    // MARK: Usage

    /// On success `limits` is non-empty and `raw` holds the JSON that produced it. On failure
    /// `limits` is empty and `raw` holds a diagnostic listing every endpoint tried with its status,
    /// so a moved/auth-failed endpoint is visible instead of silently showing zeros.
    struct UsageResult { let limits: [UsageLimit]; let raw: String? }

    func getUsage(orgId: String) async -> UsageResult {
        // claude.ai serves usage from /usage as percent-per-window limit objects.
        // Keep fallbacks in case the path moves.
        let endpoints = [
            "/api/organizations/\(orgId)/usage",
            "/api/organizations/\(orgId)/limits",
            "/api/account/usage",
            "/api/usage"
        ]

        var diagnostics = "No usage endpoint returned usable data.\n\nTried:\n"
        for endpoint in endpoints {
            let result = await get(endpoint)
            let statusLabel = result.status == -1 ? "network error" : "HTTP \(result.status)"
            diagnostics += "\u{2022} \(endpoint) \u{2192} \(statusLabel)\n"

            guard let raw = result.body else { continue }
            let limits = UsageData.parseLimits(raw)
            if !limits.isEmpty { return UsageResult(limits: limits, raw: raw) }
        }
        return UsageResult(limits: [], raw: diagnostics)
    }

    // MARK: One-shot fetch

    enum FetchError: LocalizedError {
        case notLoggedIn
        case noOrganizations
        case noUsage(String)

        var errorDescription: String? {
            switch self {
            case .notLoggedIn: return "Not signed in."
            case .noOrganizations: return "No organizations found — session may have expired."
            case .noUsage(let diagnostic): return diagnostic
            }
        }
    }

    /// Full fetch pipeline shared by the app, the widget, and the background task.
    /// Returns fresh `UsageData` or throws a descriptive error (never blanks cached data).
    func fetchUsageData() async throws -> UsageData {
        let orgs = await getOrganizations()
        guard let org = orgs.first else { throw FetchError.noOrganizations }

        let result = await getUsage(orgId: org.effectiveId)
        guard !result.limits.isEmpty else {
            throw FetchError.noUsage(result.raw ?? "Claude returned no usage data.")
        }

        return UsageData(
            limits: result.limits,
            planName: org.planTier ?? org.billingType,
            orgName: org.name,
            lastFetchedEpoch: Date().timeIntervalSince1970,
            rawJson: result.raw
        )
    }
}
